### PR TITLE
Support `grouping` function for `roll up`

### LINKED ISSUE #7134: Support `grouping` function for `roll up`
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #7134

Problem Summary:

### What is changed and how it works?

Support the `grouping` function and the passing of meta data for function.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by xzhangxian1008** [comment]
> ### What problem does this PR solve?
> Issue Number: close #7134
> 
> Problem Summary:
> 
> ### What is changed and how it works?
> ### Check List
> Tests
> 
> * [ ]  Unit test
> * [ ]  Integration test
> * [ ]  Manual test (add detailed scripts or steps below)
> * [ ]  No code
> 
> Side effects
> 
> * [ ]  Performance regression: Consumes more CPU
> * [ ]  Performance regression: Consumes more Memory
> * [ ]  Breaking backward compatibility
> 
> Documentation
> 
> * [ ]  Affects user behaviors
> * [ ]  Contains syntax changes
> * [ ]  Contains variable changes
> * [ ]  Contains experimental features
> * [ ]  Changes MySQL compatibility
> 
> ### Release note
> ```
> None
> ```

File: dbms/src/Functions/IFunction.h [hunk 1]
```diff
@@ -173,7 +174,12 @@ class IFunctionBase
       */
     virtual Monotonicity getMonotonicityForRange(const IDataType & /*type*/, const Field & /*left*/, const Field & /*right*/) const
     {
-        throw Exception("Function " + getName() + " has no information about its monotonicity.", ErrorCodes::NOT_IMPLEMENTED);
+        throw Exception(fmt::format("Function {} has no information about its monotonicity.", getName()), ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    virtual void setMetaData(const tipb::Expr & /*expr*/)
```
Comment (yibin87) [inline dbms/src/Functions/IFunction.h:180]: Add some comments here?

File: dbms/src/Interpreters/ExpressionActions.h [hunk 2]
```diff
@@ -162,11 +163,14 @@ class ExpressionActions
...
     void addInput(const NameAndTypePair & column);
 
-    void add(const ExpressionAction & action);
+    void add(const ExpressionAction & action, const tipb::Expr * expr = nullptr);
 
     /// Adds new column names to out_new_columns (formed as a result of the added action).
     void add(const ExpressionAction & action, Names & out_new_columns);
 
+    // Set meta data for the last ExpressionAction
+    void setMetaData(const tipb::Expr & expr);
```
Comment (yibin87) [inline dbms/src/Interpreters/ExpressionActions.h:172]: when will this function be called?

File: dbms/src/Functions/IFunction.h (hunk 1 above)
Comment (xzhangxian1008) [reply dbms/src/Functions/IFunction.h:180]: > Add some comments here?

okk

File: dbms/src/Interpreters/ExpressionActions.h (hunk 2 above)
Comment (xzhangxian1008) [reply dbms/src/Interpreters/ExpressionActions.h:172]: > when will this function be called?

https://github.com/xzhangxian1008/tiflash/blob/5fd73302a69f009c47c43d30a776e5f4b7293e68/dbms/src/Functions/IFunction.h#L404-L405

This function will be called when the called tidb function needs meta data.

Comment (xzhangxian1008) [reply dbms/src/Interpreters/ExpressionActions.h:172]: > when will this function be called?

I will delete useless functions.

File: dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp [hunk 3]
```diff
@@ -666,14 +666,15 @@ String DAGExpressionAnalyzer::applyFunction(
...
     const ExpressionActionsPtr & actions,
-    const TiDB::TiDBCollatorPtr & collator)
+    const TiDB::TiDBCollatorPtr & collator,
+    const tipb::Expr * expr)
 {
     String result_name = genFuncString(func_name, arg_names, {collator});
     if (actions->getSampleBlock().has(result_name))
         return result_name;
     const FunctionBuilderPtr & function_builder = FunctionFactory::instance().get(func_name, context);
     const ExpressionAction & action = ExpressionAction::applyFunction(function_builder, arg_names, result_name, collator);
```
Comment (yibin87) [inline dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp:676]: Please check if grouping can use the instance from FunctionFactory directly

File: dbms/src/Functions/FunctionsGrouping.h [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+#include <Common/typeid_cast.h>
+#include <Core/ColumnNumbers.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+
+#include "Columns/ColumnNullable.h"
+#include "common/types.h"
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:23]: > Format include files to <xxxx/xxxx.h>

okk

File: dbms/src/Functions/FunctionsGrouping.h [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    {
+        const auto & meta_data = expr.groupingmeta();
+        version = meta_data.version();
+        size_t num = meta_data.grouping_ids_size();
+
+        if (num <= 0)
+            throw Exception("number of grouping_ids should be greater than 0");
+
+        if (version == 1 || version == 2)
+            meta_grouping_id = meta_data.grouping_ids().at(0);
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:107]: > change .at => [], "at" adds extra bound check logic which is uesless here.

okk

File: dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp (hunk 3 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp:676]: > Please check if grouping can use the instance from FunctionFactory directly

https://github.com/pingcap/tiflash/blob/aa9174f6c54ebc428e601b2611680817803fbcea/dbms/src/Functions/FunctionFactory.h#L55-L62
According to the above codes, we will create new builder each time we want to get a builder.

Comment (xzhangxian1008) [reply dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp:676]: > Please check if grouping can use the instance from FunctionFactory directly

https://github.com/pingcap/tiflash/blob/master/dbms/src/Functions/FunctionFactory.h#L55-L62
According to the above codes, new function will be created each time we want to get a function.

Comment (yibin87) [reply dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp:676]: Make sense.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        const auto * grouping_col_vec = checkAndGetColumn<ColumnVector<ArgType>>(&(*col_grouping_ids));
+        if (grouping_col_vec == nullptr)
+            throw Exception("Arg's data type should be UInt64 in grouping function.");
+
+        const typename ColumnVector<ArgType>::Container & grouping_container = grouping_col_vec->getData();
+
+        // get result's data container
+        auto col_vec_res = ColumnVector<ResultType>::create();
+        typename ColumnVector<ResultType>::Container & vec_res = col_vec_res->getData();
+        vec_res.assign(row_num, static_cast<ResultType>(0));
```
Comment (yibin87) [inline dbms/src/Functions/FunctionsGrouping.h:162]: use resize_fill instead of assign?

File: dbms/src/Functions/FunctionsGrouping.h [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+        for (size_t i = 0; i < row_num; ++i)
+        {
+            if constexpr (version == 1)
+                vec_res[i] = groupingImplV1(grouping_container[i], getMetaGroupingID());
```
Comment (yibin87) [inline dbms/src/Functions/FunctionsGrouping.h:167]: Might use SIMD tech to improve performance for V1 and V2? Just a suggestion.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+{
+public:
+    using ResultType = UInt8;
+    using ArgType = UInt64; // arg type should always be UInt64
+    static constexpr auto name = Name::name;
+
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionGrouping>(); }
+    String getName() const override { return name; }
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override { collator = collator_; }
+    bool useDefaultImplementationForNulls() const override { return false; }
```
Comment (yibin87) [inline dbms/src/Functions/FunctionsGrouping.h:49]: Can you breifly describe how the function deals with null values?

File: dbms/src/Functions/FunctionsGrouping.h [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/typeid_cast.h>
+#include <Core/ColumnNumbers.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Columns/ColumnNullable.h>
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:21]: > Format the head files, it will sort the names.

okk

File: dbms/src/Functions/FunctionsGrouping.h (hunk 6 above)
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:162]: > resize_fill

okk

File: dbms/src/Functions/FunctionsGrouping.h (hunk 8 above)
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:49]: > Can you breifly describe how the function deals with null values?

`useDefaultImplementationForNulls` is used for getting the nested column in a nullable column when it's true. However, this function is impossible to encounter a nullable column so it's needless to check nullable before entering this function.

**Review by yibin87 — COMMENTED** [review]
Others LGTM

File: dbms/src/Functions/FunctionsGrouping.h [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        {
+            if constexpr (version == 1)
+                vec_res[i] = groupingImplV1(grouping_container[i], getMetaGroupingID());
+            else if constexpr (version == 2)
+                vec_res[i] = groupingImplV2(grouping_container[i], getMetaGroupingID());
+            else if constexpr (version == 3)
+                vec_res[i] = groupingImplV3(grouping_container[i], getMetaGroupingIDs());
+            else
+                throw Exception("Invalid version in grouping function");
+        }
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:174]: > Why not call grouping(UInt64 grouping_id)?

It's some design placement before the implemention of meta. I will delete them.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+    UInt32 getVersion() const
+    {
+        return version;
+    }
+
+    UInt64 getMetaGroupingID() const
+    {
+        return meta_grouping_id;
+    }
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:217]: > Not necessary?

I will delete them.

**Review by ywqzzy — APPROVED** [review]

File: dbms/src/Functions/FunctionsGrouping.h [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+            throw Exception("Grouping function decodes meta data fail");
+
+        mode = static_cast<tipb::GroupingMode>(meta.mode());
+        size_t num = meta.grouping_marks_size();
+
+        if (num <= 0)
+            throw Exception("number of grouping_ids should be greater than 0");
+
+        if (mode == 1 || mode == 2)
+            meta_grouping_id = meta.grouping_marks()[0];
```
[the remark this answers is not in the document]
Comment (windtalker) [reply dbms/src/Functions/FunctionsGrouping.h:62]: And assert `meta_grouping_id` is the power of 2 if the mode is `ModeBitAnd`?

File: dbms/src/Functions/FunctionsGrouping.h [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+        return iter == meta_grouping_marks.end();
+    }
+
+private:
+    DataTypes argument_types;
+    DataTypePtr return_type;
+
+    tipb::GroupingMode mode;
+    UInt64 meta_grouping_id = 0;
+    std::set<UInt64> meta_grouping_marks = {};
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsGrouping.h:198]: Why use `std::set` instead of `std::unordered_set`?

File: dbms/src/Functions/IFunction.h [hunk 14]
```diff
@@ -294,10 +295,15 @@ class IFunction
 
     virtual void getLambdaArgumentTypes(DataTypes & /*arguments*/) const
     {
-        throw Exception("Function " + getName() + " can't have lambda-expressions as arguments", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        throw Exception(fmt::format("Function {} can't have lambda-expressions as arguments", getName()), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
     }
 
     virtual void setCollator(const TiDB::TiDBCollatorPtr &) {}
+
+    virtual void setMetaData(const tipb::Expr & /**/)
```
Comment (windtalker) [inline dbms/src/Functions/IFunction.h:303]: Not needed any more?

File: dbms/src/Functions/IFunction.h [hunk 15]
```diff
@@ -355,6 +361,11 @@ class DefaultFunctionBase final : public IFunctionBase
         return function->getMonotonicityForRange(type, left, right);
     }
 
+    void setMetaData(const tipb::Expr & expr)
```
Comment (windtalker) [inline dbms/src/Functions/IFunction.h:364]: Ditto

File: dbms/src/TestUtils/FunctionTestUtils.h [hunk 16]
```diff
@@ -807,7 +811,8 @@ class FunctionTest : public ::testing::Test
         const ColumnNumbers & argument_column_numbers,
         const ColumnsWithTypeAndName & columns,
         const TiDB::TiDBCollatorPtr & collator = nullptr,
-        bool raw_function_test = false);
+        bool raw_function_test = false,
```
Comment (windtalker) [inline dbms/src/TestUtils/FunctionTestUtils.h:814]: Are these changes still needed?

File: dbms/src/Flash/Coprocessor/DAGExpressionAnalyzerHelper.cpp [hunk 17]
```diff
@@ -401,6 +402,32 @@ String DAGExpressionAnalyzerHelper::buildRegexpFunction(
...
+        String name = analyzer->getActions(child, actions);
+        argument_names.push_back(name);
+    }
+
+    String result_name = genFuncString(func_name, argument_names, {getCollatorFromExpr(expr)});
+    if (actions->getSampleBlock().has(result_name))
+        return result_name;
+
+    FunctionBuilderPtr function_builder = FunctionFactory::instance().get(func_name, analyzer->getContext());
+    auto * function_builder_grouping_cast = dynamic_cast<FunctionBuilderGrouping *>(function_builder.get());
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Flash/Coprocessor/DAGExpressionAnalyzerHelper.cpp:423]: > Why have `_cast` suffix?

typo, I will fix it

File: dbms/src/Flash/Coprocessor/DAGUtils.cpp [hunk 18]
```diff
@@ -69,618 +69,618 @@ const std::unordered_map<tipb::ExprType, String> distinct_agg_func_map({
...
-    {tipb::ScalarFuncSig::Trim2Args, "tidbTrim"},
-    {tipb::ScalarFuncSig::Trim3Args, "tidbTrim"},
-    {tipb::ScalarFuncSig::LTrim, "tidbLTrim"},
-    {tipb::ScalarFuncSig::RTrim, "tidbRTrim"},
-    {tipb::ScalarFuncSig::UnHex, "tidbUnHex"},
-    {tipb::ScalarFuncSig::UpperUTF8, "upperUTF8"},
-    {tipb::ScalarFuncSig::Upper, "upperBinary"},
-    //{tipb::ScalarFuncSig::CharLength, "upper"},
-});
+const std::unordered_map<tipb::ScalarFuncSig, String> scalar_func_map({{tipb::ScalarFuncSig::CastIntAsInt, "tidb_cast"},
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Flash/Coprocessor/DAGUtils.cpp:72]: > Why change so many lines?

I don't know why, these changes are made by clang-format. Maybe someone modify the configuration of clang-format.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+        if (!meta.ParseFromString(expr.val()))
+            throw Exception("Grouping function decodes meta data fail");
+
+        mode = static_cast<tipb::GroupingMode>(meta.mode());
+        size_t num = meta.grouping_marks_size();
+
+        if (num <= 0)
+            throw Exception("number of grouping_ids should be greater than 0");
+
+        if (mode == 1 || mode == 2)
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:61]: > use enum value instead of `1` and `2`?

okk

File: dbms/src/Functions/FunctionsGrouping.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:62]: > assert num is 1 there?

okk

File: dbms/src/Functions/FunctionsGrouping.h [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+    const DataTypePtr & getReturnType() const override { return return_type; }
+
+    ExecutableFunctionPtr prepare(const Block & /*sample_block*/) const override
+    {
+        return std::const_pointer_cast<FunctionGrouping>(shared_from_this());
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        NullPresence null_presence = getNullPresense(block, arguments);
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:81]: > Why can't useDefulatImplForNulls and useDefaultImplForConstant?

I think we can use `useDefaultImplForConstant`, however, we'd better to handle nullable column in the function so that we can check the validation that no nullable column should be received in `grouping` function.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+        UInt64 grouping_id = col_grouping_ids_const->getUInt(0);
+        auto res = grouping(grouping_id);
+
+        col_res = DataTypeNumber<ResultType>().createColumnConst(row_num, Field(static_cast<UInt64>(res)));
+    }
+
+    void processNonConstGroupingIDs(const ColumnPtr & col_grouping_ids, ColumnPtr & col_res, size_t row_num) const
+    {
+        if (col_grouping_ids->isColumnNullable())
+            throw Exception("Grouping function shouldn't get nullable column");
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:109]: > Actually I'm not sure for that, and I think if we set `useDefaultImplForNulls` to true, we don't need to handle this?

No nullable column should be sent into `grouping` function, so I think we need to check this validation,

[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:109]: > > Actually I'm not sure for that, and I think if we set `useDefaultImplForNulls` to true, we don't need to handle this?
> 
> No nullable column should be sent into `grouping` function, so I think we need to check this logic,

File: dbms/src/Functions/FunctionsGrouping.h [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+            break;
+        case tipb::GroupingMode::ModeNumericSet:
+            groupingVec<3>(col_grouping_ids, col_res, row_num);
+            break;
+        default:
+            throw Exception(fmt::format("Invalid version {} in grouping function", magic_enum::enum_name(mode)));
+        };
+    }
+
+    template <int version>
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:132]: > Why not use enum here?

I will fix it.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+        }
+    }
+
+    String getName() const override { return name; }
+    const DataTypes & getArgumentTypes() const override { return argument_types; }
+    const DataTypePtr & getReturnType() const override { return return_type; }
+
+    ExecutableFunctionPtr prepare(const Block & /*sample_block*/) const override
+    {
+        return std::const_pointer_cast<FunctionGrouping>(shared_from_this());
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:76]: > just use `DefaultExecutable` is not enough?

I think it's enough.

File: dbms/src/Functions/FunctionsGrouping.h (hunk 13 above)
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:198]: > Why use `std::set` instead of `std::unordered_set`?

According to [here](https://stackoverflow.com/questions/1349734/why-would-anyone-use-set-instead-of-unordered-set#:~:text=For%20a%20small%20number%20of%20elements%2C%20lookups%20in%20a%20set%20might%20be%20faster%20than%20lookups%20in%20an%20unordered_set.), I think `set` is faster than `unordered_set`.

File: dbms/src/TestUtils/FunctionTestUtils.h (hunk 16 above)
Comment (xzhangxian1008) [reply dbms/src/TestUtils/FunctionTestUtils.h:814]: > Are these changes still needed?

Yes, we still need them to pass meta data in ut.

File: dbms/src/Functions/FunctionsGrouping.h (hunk 13 above)
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:217]: > 

okk

File: dbms/src/Functions/FunctionsGrouping.h [hunk 24]
```diff
@@ -34,6 +34,23 @@ extern const int TOO_LESS_ARGUMENTS_FOR_FUNCTION;
 extern const int TOO_MANY_ARGUMENTS_FOR_FUNCTION;
 } // namespace ErrorCodes
 
+static bool isPowerOf2(uint64_t num)
+{
+    uint64_t flag = 1ul << 63ul;
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:39]: > Why not just use
> 
> ```
> return num & (num - 1) == 0;
> ```

okk

File: dbms/src/Flash/Coprocessor/DAGExpressionAnalyzerHelper.h [hunk 25]
```diff
@@ -87,13 +87,19 @@ class DAGExpressionAnalyzerHelper
...
+        const ExpressionActionsPtr & actions);
+
     static String buildDefaultFunction(
         DAGExpressionAnalyzer * analyzer,
         const tipb::Expr & expr,
         const ExpressionActionsPtr & actions);
 
     using FunctionBuilder = std::function<String(DAGExpressionAnalyzer *, const tipb::Expr &, const ExpressionActionsPtr &)>;
     using FunctionBuilderMap = std::unordered_map<String, FunctionBuilder>;
+    using FunctionNeedMetaData = std::set<String>;
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Flash/Coprocessor/DAGExpressionAnalyzerHelper.h:102]: > Seems not used?

I will delete it

File: dbms/src/Functions/FunctionsGrouping.h [hunk 26]
```diff
@@ -0,0 +1,287 @@
...
+                vec_res[i] = groupingImplModeNumericCmp(grouping_container[i]);
+            else if constexpr (mode == tipb::GroupingMode::ModeNumericSet)
+                vec_res[i] = groupingImplModeNumericSet(grouping_container[i]);
+            else
+                throw Exception("Invalid mode in grouping function");
+        }
+        col_res = std::move(col_vec_res);
+    }
+
+    ResultType grouping(UInt64 grouping_id) const
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:180]: > Is this function still needed?

It seems not need any more, I will delete it.

File: dbms/src/Interpreters/ExpressionActions.h [hunk 27]
```diff
@@ -37,10 +39,7 @@ using NamesWithAliases = std::vector<NameWithAlias>;
 class Join;
 class Expand;
 
-class IFunctionBase;
 using FunctionBasePtr = std::shared_ptr<IFunctionBase>;
-
-class IFunctionBuilder;
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Interpreters/ExpressionActions.h:43]: > The changes in this file can are also not needed anymore?

okk, I will revoke these changes

File: dbms/src/Functions/FunctionsGrouping.h [hunk 28]
```diff
@@ -0,0 +1,287 @@
...
+    {
+        return std::const_pointer_cast<FunctionGrouping>(shared_from_this());
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        NullPresence null_presence = getNullPresense(block, arguments);
+        if (null_presence.has_null_constant)
+        {
+            block.getByPosition(result).column = block.getByPosition(result).type->createColumnConst(block.rows(), Null());
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:108]: > I'm not very understand the logic: Nullable columns is not allowed, but null-constant is allowed? I suggest just set `useDefaultImplForNulls ` to true and remove all the null related code in this function.

okk

File: dbms/src/TestUtils/FunctionTestUtils.h [hunk 29]
```diff
@@ -816,6 +823,21 @@ class FunctionTest : public ::testing::Test
         return executeFunction(func_name, argument_column_numbers, vec);
     }
 
+    ColumnWithTypeAndName executeFunctionWithMetaData(
+        const String & func_name,
+        const ColumnsWithTypeAndName & columns,
+        const TiDB::TiDBCollatorPtr & collator = nullptr,
+        const FuncMetaData & meta = FuncMetaData(),
+        bool raw_function_test = false);
```
Comment (windtalker) [inline dbms/src/TestUtils/FunctionTestUtils.h:831]: I think here is no need to add `raw_function_test ` for `executeFunctionWithMetaData ` because function with meta can't execute in raw_function_test mode, and also I think since `executeFunctionWithMetaData ` is a new function that should only be used for function that has function meta, there is no need to set default value of `meta`

File: dbms/src/TestUtils/FunctionTestUtils.h [hunk 30]
```diff
@@ -816,6 +823,21 @@ class FunctionTest : public ::testing::Test
...
+
+    ColumnWithTypeAndName executeFunctionWithMetaData(
```
Comment (windtalker) [inline dbms/src/TestUtils/FunctionTestUtils.h:833]: Ditto.

File: dbms/src/TestUtils/ColumnsToTiPBExpr.cpp [hunk 31]
```diff
@@ -179,6 +181,9 @@ void columnsToTiPBExpr(
     }
     else
     {
+        if (func_name == "grouping")
```
Comment (windtalker) [inline dbms/src/TestUtils/ColumnsToTiPBExpr.cpp:184]: I think `expr->set_val(val)` can always be called if `val` is not empty.

File: dbms/src/TestUtils/FunctionTestUtils.cpp [hunk 32]
```diff
@@ -403,7 +407,7 @@ DataTypePtr getReturnTypeForFunction(
         for (size_t i = 0; i < columns.size(); ++i)
             argument_column_numbers.push_back(i);
         auto columns_with_unique_name = toColumnsWithUniqueName(columns);
-        auto [actions, result_name] = buildFunction(context, func_name, argument_column_numbers, columns_with_unique_name, collator);
+        auto [actions, result_name] = buildFunction(context, func_name, argument_column_numbers, columns_with_unique_name, collator, "");
```
Comment (windtalker) [inline dbms/src/TestUtils/FunctionTestUtils.cpp:410]: Why change this? The last argument should have default value

File: dbms/src/TestUtils/ColumnsToTiPBExpr.h [hunk 33]
```diff
@@ -20,6 +20,7 @@
 #include <Core/Field.h>
 #include <Core/Types.h>
 #include <DataTypes/IDataType.h>
+#include <tipb/expression.pb.h>
```
Comment (windtalker) [inline dbms/src/TestUtils/ColumnsToTiPBExpr.h:23]: Is this file still needed to be include?

File: dbms/src/Functions/FunctionsGrouping.h [hunk 34]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        if (!context.getDAGContext())
+        {
+            throw Exception("DAGContext should not be nullptr.", ErrorCodes::LOGICAL_ERROR);
+        }
+        return std::make_shared<FunctionBuilderGrouping>(context);
+    }
+
+    String getName() const override { return name; }
+    bool useDefaultImplementationForNulls() const override { return false; }
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:191]: > should return `true` here?

yes, i will fix.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 35]
```diff
@@ -0,0 +1,241 @@
...
+        else if (arg_num > 1)
+            throw Exception("Too many arguments", ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION);
+
+        const DataTypePtr & arg_data_type = arguments[0].type;
+        if (arg_data_type->isNullable())
+        {
+            const auto * null_type = checkAndGetDataType<DataTypeNullable>(arg_data_type.get());
+            assert(null_type != nullptr);
+
+            const auto & nested_type = null_type->getNestedType();
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:225]: > `const auto & nested_type = removeNullable(arg_data_type)` should be enough.

okk

File: dbms/src/Functions/FunctionsGrouping.h [hunk 36]
```diff
@@ -0,0 +1,241 @@
...
+            if (nested_type->isInteger())
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsGrouping.h:226]: > maybe just check the type is `UInt64`?

okk, I will check.

File: dbms/src/TestUtils/ColumnsToTiPBExpr.cpp (hunk 31 above)
Comment (xzhangxian1008) [reply dbms/src/TestUtils/ColumnsToTiPBExpr.cpp:184]: > I think `expr->set_val(val)` can always be called if `val` is not empty.

okk, I will always call `expr->set_val(val)`

File: dbms/src/TestUtils/FunctionTestUtils.h (hunk 29 above)
Comment (xzhangxian1008) [reply dbms/src/TestUtils/FunctionTestUtils.h:831]: > I think here is no need to add `raw_function_test ` for `executeFunctionWithMetaData ` because function with meta can't execute in raw_function_test mode, and also I think since `executeFunctionWithMetaData ` is a new function that should only be used for function that has function meta, there is no need to set default value of `meta`

I will change it

File: dbms/src/TestUtils/FunctionTestUtils.cpp (hunk 32 above)
Comment (xzhangxian1008) [reply dbms/src/TestUtils/FunctionTestUtils.cpp:410]: > Why change this? The last argument should have default value

val of buildFunction does not have default value, so I need to explicitly designate it.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 37]
```diff
@@ -0,0 +1,245 @@
...
+        if (arg_num < 1)
+            throw Exception("Too few arguments", ErrorCodes::TOO_LESS_ARGUMENTS_FOR_FUNCTION);
+        else if (arg_num > 1)
+            throw Exception("Too many arguments", ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION);
+
+        const DataTypePtr & arg_data_type = arguments[0].type;
+        if (arg_data_type->isNullable())
+        {
+            const auto * null_type = checkAndGetDataType<DataTypeNullable>(arg_data_type.get());
+            assert(null_type != nullptr);
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsGrouping.h:225]: Above 2 line seems useless.
